v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/endtoend/ForwardAndReverseMigrationEndToEndIT.java

```diff
@@ -0,0 +1,222 @@
+  public void setUp() throws IOException, InterruptedException {
+    skipBaseCleanup = true;
+    synchronized (ForwardAndReverseMigrationEndToEndIT.class) {
+      testInstances.add(this);
+      if (rrJobInfo == null || fwdJobInfo == null) {
+        // create Spanner Resources
+        spannerResourceManager =
+            createSpannerDatabase(ForwardAndReverseMigrationEndToEndIT.SPANNER_DDL_RESOURCE);
+        spannerMetadataResourceManager = createSpannerMetadataDatabase();
+
+        // fetch secrets
+        secretClient = SecretManagerResourceManager.builder(PROJECT, credentialsProvider).build();
+        // create MySql Resources
+        cloudSqlResourceManager = CloudMySQLResourceManager.builder(testName).build();
+        jdbcSource =
+            createMySqlDatabase(
+                cloudSqlResourceManager,
+                new HashMap<>() {
+                  {
+                    put(TABLE, AUTHOR_TABLE_COLUMNS);
+                  }
+                });
+
+        // create and upload GCS Resources
+        gcsResourceManager =
+            GcsResourceManager.builder(artifactBucketName, getClass().getSimpleName(), credentials)
+                .build();
+        createAndUploadReverseShardConfigToGcs(
+            gcsResourceManager, cloudSqlResourceManager, cloudSqlResourceManager.getHost());
+        gcsResourceManager.createArtifact(
+            "input/session.json",
+            generateSessionFile(
+                cloudSqlResourceManager.getDatabaseName(),
+                spannerResourceManager.getDatabaseId(),
+                SESSION_FILE_RESOURCE));
+
+        // create pubsub manager
+        pubsubResourceManager = setUpPubSubResourceManager();
+
+        // launch forward migration template
+        fwdJobInfo =
+            launchFwdDataflowJob(spannerResourceManager, gcsResourceManager, pubsubResourceManager);
+
+        // launch reverse migration template
+        rrJobInfo =
+            launchRRDataflowJob(
+                spannerResourceManager,
+                gcsResourceManager,
+                spannerMetadataResourceManager,
+                pubsubResourceManager,
+                MYSQL_SOURCE_TYPE);
+      }
+    }
+  }
+
+  /**
+   * Cleanup dataflow job and all the resources and resource managers.
+   *
+   * @throws IOException
+   */
+  @AfterClass
+  public static void cleanUp() throws IOException {
+    for (ForwardAndReverseMigrationEndToEndIT instance : testInstances) {
+      instance.tearDownBase();
+    }
+    ResourceManagerUtils.cleanResources(
+        spannerResourceManager,
+        spannerMetadataResourceManager,
+        gcsResourceManager,
+        pubsubResourceManager,
+        cloudSqlResourceManager);
+  }
+
+  @Test
+  public void spannerToSourceDbBasic() {
+    // Forward Migration check condition
+    assertThatPipeline(fwdJobInfo).isRunning();
+    writeRowInMySqlAndAssertRows();
+
+    // Reverse Migration check condition
+    assertThatPipeline(rrJobInfo).isRunning();
```

Comment:
Lets assert that both the pipelines are running before inserting any rows In MySql Or Spanner. That is,

1. assert fwd job is running.
2. assert rr job is running .
3. writeRowInMySqlAndAssertRows
4. writeRowInSpanner ...